<a href="https://colab.research.google.com/github/cam31415/ML-project-COLAS-/blob/main/ML_Project_COLAS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Machine Learning Project**
For the explanation of the code cf. the report 

#MLP model 

**Step 1 :** Import the necessary librairies 

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import tensorflow as tf
from keras.optimizers import SGD, RMSprop
import numpy as np


**Step 2 :** Load the data 

In [ ]:
x_train_url = 'https://raw.githubusercontent.com/cam31415/ML-project-COLAS-/main/x_train.csv'
y_train_url = 'https://raw.githubusercontent.com/cam31415/ML-project-COLAS-/main/y_train.csv'
x_test_url = 'https://raw.githubusercontent.com/cam31415/ML-project-COLAS-/main/x_test.csv'
df_X_train = pd.read_csv(x_train_url)
df_Y_train = pd.read_csv(y_train_url)
df_X_test = pd.read_csv(x_test_url)

**Step 3 :** Preprocess the data 

In [ ]:
df_X_train.drop('Tmix', inplace=True, axis=1)
df_X_train.drop('m_power', inplace=True, axis=1)
df_X_train.drop('Tosc', inplace=True, axis=1)
df_X_train.drop('Unnamed: 0', inplace=True, axis=1)

df_X_test.drop('Tmix', inplace=True, axis=1)
df_X_test.drop('m_power', inplace=True, axis=1)
df_X_test.drop('Tosc', inplace=True, axis=1)
df_X_test.drop('Unnamed: 0', inplace=True, axis=1)

df_Y_train.drop('Unnamed: 0', inplace=True, axis=1)
df_Y_train['target'] = df_Y_train['target']-1 

In [ ]:
df_X_train 


cfo_demod  gain_imb     iq_imb     or_off  quadr_err    ph_err  \
0      592.234802  0.048079 -35.082729 -28.560846   1.993170  1.107926   
1     -103.302032 -0.019917 -29.946953 -35.798664  -3.642311  1.236059   
2     -582.331299  0.036379 -32.096672 -31.905628   2.835839  1.282163   
3     -630.611267  0.063928 -38.216297 -30.084171   1.346316  1.154848   
4     -415.526978 -0.055761 -29.180740 -36.601025  -3.963526  1.498889   
...           ...       ...        ...        ...        ...       ...   
15355 -476.575653  0.039874 -39.792461 -29.962997   1.143736  1.175517   
15356 -345.645508  0.077925 -32.963398 -27.998444   2.524010  1.159356   
15357 -613.989807  0.090771 -38.951530 -28.111988   1.145893  0.995857   
15358 -997.769531  0.099840 -39.532818 -27.819826   1.014136  1.069736   
15359  450.446838  0.068571 -34.959385 -28.346176   1.996514  1.035648   

        mag_err       evm  
0      1.507550  2.423943  
1      2.741568  3.458056  
2      2.140096  3.013522  
3      1.093465  2.254514  
4      3.608737  4.286684  
...         ...       ...  
15355  1.044636  2.224311  
15356  1.948950  2.761400  
15357  0.938308  1.939818  
15358  0.875986  2.026062  
15359  1.527475  2.357938  

[15360 rows x 8 columns]

In [ ]:
df_X_test

cfo_demod  gain_imb     iq_imb     or_off  quadr_err    ph_err  \
0      516.312866  0.055910 -34.748932 -28.086121   2.064618  1.166067   
1     -452.875031  0.091061 -31.916044 -29.092400   2.843033  1.242625   
2      543.581909  0.067271 -34.311523 -28.395725   2.160502  1.031588   
3     -711.415710  0.001931 -30.205307 -35.895462  -3.537908  1.260278   
4     -912.994873  0.054695 -31.485973 -33.110847   3.031799  1.343741   
...           ...       ...        ...        ...        ...       ...   
3835 -1129.111816  0.071651 -40.088821 -27.627096   1.045999  0.959680   
3836  -201.570984 -0.013889 -30.041744 -36.312626  -3.603977  1.277607   
3837  -434.601959 -0.067320 -29.236820 -37.041874  -3.929963  1.581971   
3838  -507.450348  0.106392 -35.374168 -29.681749   1.821150  1.218159   
3839   -70.713585 -0.024844 -29.677410 -35.998009  -3.758281  1.339186   

       mag_err       evm  
0     1.528550  2.541858  
1     2.044024  2.974349  
2     1.615977  2.427548  
3     2.586169  3.350533  
4     2.222488  3.199810  
...        ...       ...  
3835  0.863437  1.836749  
3836  2.666203  3.413867  
3837  3.663174  4.409636  
3838  1.373915  2.490288  
3839  2.719270  3.557638  

[3840 rows x 8 columns]

In [ ]:
df_Y_train

target
0           4
1           0
2           5
3           2
4           1
...       ...
15355       2
15356       7
15357       6
15358       6
15359       4

[15360 rows x 1 columns]

**Step 4 :** Scale the features 

In [ ]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df_X_train)
df_X_test = scaler.fit_transform(df_X_test)


**Step 5 :** Split the data into training and test sets 

In [ ]:
df_X_train, X_test_other, df_Y_train, y_test = train_test_split(X_scaled, df_Y_train, test_size=0.25, random_state=42)
print("y_train shape", y_test.shape)

y_train shape (3840, 1)


We chose a test size that is 25% of the total size because we have df_X_test that is 3840 columns and 25% of 15360 is 3840, so that the sizes match 

**Step 6 :** Define the optimizer, build  and compile the mlp model 

In [ ]:
optimizer = tf.keras.optimizers.SGD(learning_rate=0.7)

model = Sequential()
model.add(Dense(64, input_dim=df_X_train.shape[1], activation='relu'))
model.add(Dense(128, activation='relu'))
model.add(Dense(8, activation='softmax'))


model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])


**Step 7 :** Train the model 

In [ ]:
history = model.fit(df_X_train, df_Y_train, epochs=100, batch_size=64, verbose=2, validation_data=(X_test_other, y_test))


Epoch 1/100
180/180 - 1s - loss: 0.1841 - accuracy: 0.9529 - val_loss: 0.0104 - val_accuracy: 0.9982 - 1s/epoch - 8ms/step
Epoch 2/100
180/180 - 0s - loss: 0.0091 - accuracy: 0.9984 - val_loss: 0.0088 - val_accuracy: 0.9990 - 470ms/epoch - 3ms/step
Epoch 3/100
180/180 - 0s - loss: 0.0078 - accuracy: 0.9987 - val_loss: 0.0048 - val_accuracy: 0.9990 - 495ms/epoch - 3ms/step
Epoch 4/100
180/180 - 0s - loss: 0.0057 - accuracy: 0.9989 - val_loss: 0.0039 - val_accuracy: 0.9990 - 465ms/epoch - 3ms/step
Epoch 5/100
180/180 - 0s - loss: 0.0035 - accuracy: 0.9990 - val_loss: 0.0043 - val_accuracy: 0.9990 - 458ms/epoch - 3ms/step
Epoch 6/100
180/180 - 0s - loss: 0.0044 - accuracy: 0.9990 - val_loss: 0.0022 - val_accuracy: 0.9992 - 488ms/epoch - 3ms/step
Epoch 7/100
180/180 - 0s - loss: 0.0027 - accuracy: 0.9990 - val_loss: 0.0037 - val_accuracy: 0.9990 - 494ms/epoch - 3ms/step
Epoch 8/100
180/180 - 0s - loss: 0.0025 - accuracy: 0.9992 - val_loss: 0.0025 - val_accuracy: 0.9990 - 432ms/epoch - 2ms/

**Step 8 :** Predict 

In [ ]:
y_pred_prob = model.predict(df_X_test)
y_pred = np.argmax(y_pred_prob, axis=1)+1 

120/120 [==============================] - 0s 1ms/step


**Step 9 :** Addition of the 'id' column on the test set & concatenate it and the predicted target column

In [ ]:
df_test = pd.DataFrame(df_X_test)
df_test['id'] = df_test.index

y_pred_df = pd.DataFrame(y_pred, columns=['target'])
y_pred_df['id'] = df_test['id']
y_pred_df = y_pred_df[['id', 'target']]


In [ ]:
y_pred_df

id  target
0        0       5
1        1       8
2        2       5
3        3       1
4        4       6
...    ...     ...
3835  3835       7
3836  3836       1
3837  3837       2
3838  3838       3
3839  3839       1

[3840 rows x 2 columns]

**Step 10 :** Import the CSV on drive 

In [ ]:
from google.colab import drive
y_pred_df.to_csv('MLP_predictions.csv', index=False)

drive.mount('/content/drive')

output = '/content/drive/My Drive/MLP_predictions.csv'
y_pred_df.to_csv(output, index=False)

Mounted at /content/drive


#SVM model

**Step 1 :** Import the necessary librairies 

In [ ]:
from sklearn.linear_model import Perceptron, LogisticRegression
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn import datasets
from sklearn import metrics

**Step 2 :** Load the data 

In [ ]:
x_train_url = 'https://raw.githubusercontent.com/cam31415/ML-project-COLAS-/main/x_train.csv'
y_train_url = 'https://raw.githubusercontent.com/cam31415/ML-project-COLAS-/main/y_train.csv'
x_test_url = 'https://raw.githubusercontent.com/cam31415/ML-project-COLAS-/main/x_test.csv'
df_X_train = pd.read_csv(x_train_url)
df_Y_train = pd.read_csv(y_train_url)
df_X_test = pd.read_csv(x_test_url)

**Step 3 :** Preprocess the data 

In [ ]:
df_X_train.drop('Tmix', inplace=True, axis=1)
df_X_train.drop('m_power', inplace=True, axis=1)
df_X_train.drop('Tosc', inplace=True, axis=1)
df_X_train.drop('Unnamed: 0', inplace=True, axis=1)

df_X_test.drop('Tmix', inplace=True, axis=1)
df_X_test.drop('m_power', inplace=True, axis=1)
df_X_test.drop('Tosc', inplace=True, axis=1)
df_X_test.drop('Unnamed: 0', inplace=True, axis=1)

df_Y_train.drop('Unnamed: 0', inplace=True, axis=1)

# Scale the features using StandardScaler
#scaler = StandardScaler()
#X_scaled = scaler.fit_transform(X_train)
#X_test = scaler.fit_transform(X_test)

# Split the data into training and testing sets
#X_train, X_test_other, Y_train, y_test = train_test_split(X_scaled, Y_train, test_size=0.25, random_state=28)
#print("y_train shape", y_test.shape)



**Step 4 :** Instantiate the Support Vector Classifier (SVC)

In [ ]:
svc = SVC(C=1.0, random_state=1, kernel='linear')

**Step 5 :** Train the model 

In [ ]:
svc.fit(df_X_train, df_Y_train)

/usr/local/lib/python3.9/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


SVC(kernel='linear', random_state=1)

**Step 6 :** Predict 

In [ ]:
y_predict = svc.predict(df_X_test)
# Calculate the accuracy of the model
#accuracy = accuracy_score(y_test, y_pred)
#print("Accuracy:", accuracy)


In [ ]:
y_predict_df = pd.DataFrame(y_predict)
y_predict_df

0
0     5
1     8
2     5
3     1
4     6
...  ..
3835  7
3836  1
3837  2
3838  3
3839  1

[3840 rows x 1 columns]